### PR TITLE
Enhance slice operation to support more range variation

### PR DESCRIPTION
**Pull request by antimora**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#1915, #1981, #1982, #1983

### Changes

Primarily to support ONNX slice operation and closer PyTorch compatibility.

1. Enhance slice operation to support more range variations: None for a missing rage and negative ranks in ranges.
2. Clamp range start and end to dimension ranks.

### Testing

More unit tests

### CODE REVIEW

**Comment by antimora** [comment]
@johnhuichen, please review this PR if you have a chance since you worked in this area previosly.

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 1]
```diff
@@ -579,50 +579,64 @@ where
...
-    ///     // v[0][0][0] = tensor[0][1][1];
-    ///     // v[0][1][0] = tensor[0][2][1];
-    ///     let tensor_slices = tensor.slice([0..1, 1..3, 1..2]);
-    ///     println!("\nexpecting [1, 2, 1] : {:?}", tensor_slices.dims());
-    ///     println!("expecting [[[5],[9]]] : {:?}", tensor_slices);
+    ///
+    ///     // 1D slicing
+    ///     let tensor = Tensor::<B, 1, burn_tensor::Int>::arange(0..5, &device);
+    ///     let slice = tensor.slice([1..4]);
+    ///     assert_eq!(slice.into_data().to_vec(), vec![1, 2, 3]);
```
Comment (laggui) [inline crates/burn-tensor/src/tensor/api/base.rs:613]: `to_vec()` returns a result, that's why the CI is failing. Just need to unwrap.

**Review by laggui — COMMENTED** [review]
LGTM! Just gotta fix the CI errors on the doc tests.

File: crates/burn-import/src/onnx/op_configuration.rs [hunk 2]
```diff
@@ -1283,60 +1283,74 @@ pub fn shape_config(curr: &Node) -> (usize, usize) {
...
-                    i += 1;
-                })
-            }
-            "steps" => value.clone().into_i64s().into_iter().for_each(|x| {
-                if x != 1 {
-                    panic!("Slice: steps other than 1 are not supported");
-                }
-            }),
+            "ends" => starts = value.clone().into_i64s(),
+            "starts" => ends = value.clone().into_i64s(),
```
Comment (johnhuichen) [inline crates/burn-import/src/onnx/op_configuration.rs:1310]: Looks like this doesn't handle negative start and end values any more.

Comment (johnhuichen) [reply crates/burn-import/src/onnx/op_configuration.rs:1310]: Ah I see. The rust slice method now handles negative dim. Nice

File: crates/burn-import/src/onnx/op_configuration.rs [hunk 3]
```diff
@@ -1283,60 +1283,74 @@ pub fn shape_config(curr: &Node) -> (usize, usize) {
...
+    }
+
+    // assert len(starts) == len(ends) == len(axes)
+    if starts.len() != ends.len() || starts.len() != axes.len() {
+        panic!("Slice: starts, ends, and axes must have the same length");
+    }
+
+    // If dim is negative, it is counted from the end
+    // Negative value means counting dimensions from the back.
+    for i in 0..axes.len() {
```
Comment (laggui) [inline crates/burn-import/src/onnx/op_configuration.rs:1339]: Didn't notice it before, and somehow clippy didn't run with the CI, but in the review mode it's getting triggered:

```
[clippy] reported by reviewdog 🐶
error: the loop variable `i` is only used to index `axes`
    --> crates/burn-import/src/onnx/op_configuration.rs:1339:14
     |
1339 |     for i in 0..axes.len() {
     |              ^^^^^^^^^^^^^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_range_loop
     = note: `-D clippy::needless-range-loop` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(clippy::needless_range_loop)]`
help: consider using an iterator
     |
1339 |     for <item> in &mut axes {
     |         ~~~~~~    ~~~~~~~~~
```

Fair point clippy 😄

Comment (antimora) [reply crates/burn-import/src/onnx/op_configuration.rs]: My local run-checks is broken =(

Fixed.

**Review by laggui — APPROVED** [review]

**Pull request merged**